azure-iot-device/tests/iothub/test_sync_clients.py

```diff
@@ -40,67 +62,154 @@ def test_from_authentication_provider_instantiates_client(
-        assert client._pipeline.on_transport_connected is not None
-        assert client._pipeline.on_transport_connected == client._on_state_change
 
-    def test_instantiation_sets_on_disconnected_handler_in_transport(self, client):
-        assert client._pipeline.on_transport_disconnected is not None
-        assert client._pipeline.on_transport_disconnected == client._on_state_change
+class SharedClientConnectTests(object):
+    @pytest.mark.it("Begins a 'connect' pipeline operation")
+    def test_calls_pipeline_connect(self, client, pipeline):
+        client.connect()
+        assert pipeline.connect.call_count == 1
 
-    def test_instantiation_sets_on_method_request_received_handler_in_transport(self, client):
-        assert client._pipeline.on_transport_method_request_received is not None
-        assert (
-            client._pipeline.on_transport_method_request_received
-            == client._inbox_manager.route_method_request
-        )
+    @pytest.mark.it("Waits for the completion of the 'connect' pipeline operation before returning")
+    def test_waits_for_pipeline_op_completion(self, mocker, client_manual_cb, pipeline_manual_cb):
+        client = client_manual_cb
+        pipeline = pipeline_manual_cb
+        event_init_mock = mocker.patch.object(threading, "Event")
+        event_mock = event_init_mock.return_value
 
-    def test_state_change_handler_clears_method_request_inboxes_on_disconnect(self, client, mocker):
-        clear_method_request_spy = mocker.spy(client._inbox_manager, "clear_all_method_requests")
-        client._on_state_change("disconnected")
-        assert clear_method_request_spy.call_count == 1
+        def check_callback_completes_event():
+            # Assert waiting for Event to complete
+            assert event_mock.wait.call_count == 1
+            assert event_mock.set.call_count == 0
+
+            # Manually trigger callback
+            cb = pipeline.connect.call_args[1]["callback"]
+            cb()
+
+            # Assert Event is now completed
+            assert event_mock.set.call_count == 1
+
+        event_mock.wait.side_effect = check_callback_completes_event
 
-    def test_connect_calls_transport(self, client, pipeline):
         client.connect()
-        assert pipeline.connect.call_count == 1
 
-    def test_disconnect_calls_transport(self, client, pipeline):
+
+class SharedClientDisconnectTests(object):
+    @pytest.mark.it("Begins a 'disconnect' pipeline operation")
+    def test_calls_pipeline_disconnect(self, client, pipeline):
         client.disconnect()
         assert pipeline.disconnect.call_count == 1
 
-    def test_send_event_calls_transport(self, client, pipeline):
+    @pytest.mark.it(
+        "Waits for the completion of the 'disconnect' pipeline operation before returning"
+    )
+    def test_waits_for_pipeline_op_completion(self, mocker, client_manual_cb, pipeline_manual_cb):
+        client = client_manual_cb
+        pipeline = pipeline_manual_cb
```

Comment:
There's a trick that I learned for this that I really like.  Instead of have a "client automatic" and "client manual" fixture with 2 different names, you could have a single "client" fixture that defaults to automatic callback.  Then, you can have a second fixture that changes your client fixture into manual.  This works because the fixtures have method scope, and your "convert to manual fixture" can only change the client, it can't replace it, but it's still cool.

```
@pytest.fixture
def client()
    return MockAutomaticClient()
    
@pytest.fixture
def make_client_manual(client, mocker):
    client.method_foo = mocker.MagicMock()
    client.method_bar = mocker.MagicMock()
    
def test_blah(client, make_client_manual):
    # client object is from MockAutomaticClient() in the client fixture
    # client.method_foo and method_bar were replaced by the make_client_manual fixture
```

Reply:
This is actually a very nice pattern that I will keep in mind, but I think I'll pass on using it here for the following reasons:

1) I prefer to avoid accessing convention-private attributes when possible. It's not a hard and fast rule, because it does have to happen (fairly frequently too...), but I like to avoid it where I can.
2) I like the symmetry between `pipeline_manual_cb` and `client_manual_cb`. If we were to patch the `client._transport` attribute to be `pipeline_manual_cb`, within `make_client_manual`, then the relationship between `client` and `pipeline_manual_cb` is obfuscated within the test.
3) "Well Carter, why even have `pipeline_manual_cb` when all it is, is a plain mocker.MagicMock()?" Okay, fair, I could get rid of that fixture altogether, but then how do we access the pipeline? "Through the client of course, `client._pipeline`" Okay, fair, but that's another access to a convention-private attribute. "Come on, really? You could remove TWO WHOLE FIXTURES BY DOING IT" Okay, fair, that's a... really good point... BUT... now the tests lose symmetry with all the other tests that import client and pipeline, and it becomes harder to parse why and how this test is different from the others.

Building off that last point/imagined rubber duck discussion with you, I think it would have been beneficial to consider this pattern earlier on in the overall construction of these tests, but without overhauling the fixture model used here, I think the way it currently is slightly gets the edge. I'll definitely keep it mind for future test patterns though, it's very nice.

---
In reply to: [291192784](https://github.com/Azure/azure-iot-sdk-python-preview/pull/101#discussion_r291192784) [](ancestors = 291192784)